File: src/advanced_localization/vision/optical_flow_layer.py
from __future__ import annotations

from typing import Optional

import cv2
import numpy as np
from numpy.typing import NDArray

from ..config import SystemConfig
from ..types import FlowMeasurement


class OpticalFlowLayer:

    def get_flow_norm_and_feature_count(
        self,
        frame_gray: NDArray[np.uint8],
        timestamp_s: float,
        altitude_m: float,
        yaw_rad: float,
    ) -> Optional[tuple[float, int]]:
        """
        Returns (flow_norm, feature_count) for ZUPT triggering.
        """
        result = self.run(frame_gray, timestamp_s, altitude_m, yaw_rad)
        if result is not None:
            flow_norm = np.linalg.norm(result.velocity_xy_mps)
            return flow_norm, result.feature_count
        return None

    def __init__(self, config: SystemConfig, camera_matrix: NDArray[np.float64]) -> None:
        self._config = config
        self._camera_matrix = np.asarray(camera_matrix, dtype=np.float64)

        self._prev_frame: Optional[NDArray[np.uint8]] = None
        self._prev_timestamp_s: Optional[float] = None
        self._prev_points: Optional[NDArray[np.float32]] = None

        self._lk_win_size = (21, 21)
        self._lk_max_level = 3
        self._lk_criteria = (cv2.TERM_CRITERIA_EPS | cv2.TERM_CRITERIA_COUNT, 30, 0.01)

    def run(
        self,
        frame_gray: NDArray[np.uint8],
        timestamp_s: float,
        altitude_m: float,
        yaw_rad: float,
    ) -> Optional[FlowMeasurement]:
        if self._prev_frame is None or self._prev_timestamp_s is None:
            self._bootstrap(frame_gray, timestamp_s)
            return None

        dt = float(timestamp_s - self._prev_timestamp_s)
        if dt <= 1e-6:
            return None

        if self._prev_points is None or len(self._prev_points) < self._config.flow.min_features:
            self._prev_points = self._detect_features(self._prev_frame)
            if self._prev_points is None or len(self._prev_points) < 4:
                self._bootstrap(frame_gray, timestamp_s)
                return None

        next_points, status_fw, _ = cv2.calcOpticalFlowPyrLK(
            self._prev_frame,
            frame_gray,
            self._prev_points,
            None,
            winSize=self._lk_win_size,
            maxLevel=self._lk_max_level,
            criteria=self._lk_criteria,
        )
        if next_points is None or status_fw is None:
            self._bootstrap(frame_gray, timestamp_s)
            return None

        back_points, status_bw, _ = cv2.calcOpticalFlowPyrLK(
            frame_gray,
            self._prev_frame,
            next_points,
            None,
            winSize=self._lk_win_size,
            maxLevel=self._lk_max_level,
            criteria=self._lk_criteria,
        )
        if back_points is None or status_bw is None:
            self._bootstrap(frame_gray, timestamp_s)
            return None

        fw_ok = status_fw.reshape(-1) == 1
        bw_ok = status_bw.reshape(-1) == 1
        bidirectional_err = np.linalg.norm(back_points - self._prev_points, axis=2).reshape(-1)
        valid = fw_ok & bw_ok & (bidirectional_err < self._config.flow.max_bidirectional_error_px)

        prev_valid = self._prev_points[valid][:, 0, :]
        next_valid = next_points[valid][:, 0, :]

        if prev_valid.shape[0] < self._config.flow.min_features:
            self._bootstrap(frame_gray, timestamp_s)
            return None

        flow_px = np.median(next_valid - prev_valid, axis=0)
        velocity_xy_enu_mps = self._pixel_flow_to_velocity_enu(
            flow_px=flow_px,
            dt=dt,
            altitude_m=altitude_m,
            yaw_rad=yaw_rad,
        )

        feature_scale = np.clip(self._config.flow.min_features / max(float(prev_valid.shape[0]), 1.0), 0.5, 4.0)
        sigma = self._config.flow.noise_std_mps * feature_scale
        covariance = np.zeros((2, 2), dtype=np.float64)
        covariance[0, 0] = sigma * sigma
        covariance[1, 1] = sigma * sigma

        self._prev_frame = frame_gray.copy()
        self._prev_timestamp_s = timestamp_s
        self._prev_points = next_valid.reshape(-1, 1, 2).astype(np.float32)

        return FlowMeasurement(
            timestamp_s=timestamp_s,
            velocity_xy_mps=velocity_xy_enu_mps,
            covariance_2x2=covariance,
            feature_count=int(prev_valid.shape[0]),
        )

    def _bootstrap(self, frame_gray: NDArray[np.uint8], timestamp_s: float) -> None:
        self._prev_frame = frame_gray.copy()
        self._prev_timestamp_s = timestamp_s
        self._prev_points = self._detect_features(frame_gray)

    def _detect_features(self, frame_gray: NDArray[np.uint8]) -> Optional[NDArray[np.float32]]:
        return cv2.goodFeaturesToTrack(
            frame_gray,
            maxCorners=800,
            qualityLevel=0.01,
            minDistance=6.0,
            blockSize=7,
            useHarrisDetector=False,
        )

    def _pixel_flow_to_velocity_enu(
        self,
        flow_px: NDArray[np.float32],
        dt: float,
        altitude_m: float,
        yaw_rad: float,
    ) -> NDArray[np.float64]:
        fx = float(self._camera_matrix[0, 0])
        fy = float(self._camera_matrix[1, 1])

        alt = max(1.0, abs(altitude_m))
        # Body-frame convention (shared with video testing):
        # forward velocity from v-flow, right velocity from -u-flow.
        vx_body = float(flow_px[1]) * alt / max(fy, 1e-6) / dt
        vy_body = -float(flow_px[0]) * alt / max(fx, 1e-6) / dt

        c = float(np.cos(yaw_rad))
        s = float(np.sin(yaw_rad))
        return np.array(
            [
                c * vx_body - s * vy_body,
                s * vx_body + c * vy_body,
            ],
            dtype=np.float64,
        )

File: advanced_localization_cpp/src/optical_flow_layer.cpp
#include "advanced_localization_cpp/optical_flow_layer.hpp"

#include <opencv2/imgproc.hpp>
#include <opencv2/video/tracking.hpp>

#include <algorithm>
#include <cmath>

namespace advanced_localization
{

OpticalFlowLayer::OpticalFlowLayer(const SystemConfig & config, const Matrix3 & camera_matrix)
: config_(config),
  camera_matrix_(camera_matrix)
{
}

std::optional<FlowMeasurement> OpticalFlowLayer::run(const cv::Mat & frame_gray, double timestamp_s, double altitude_m, double yaw_rad)
{
  if (frame_gray.empty()) {
    return std::nullopt;
  }
  double frame_scale = 1.0;
  const cv::Mat frame_work = prepare_frame(frame_gray, frame_scale);
  if (prev_frame_.empty() || !prev_timestamp_s_.has_value()) {
    bootstrap(frame_work, timestamp_s, frame_scale);
    return std::nullopt;
  }

  const double dt = timestamp_s - *prev_timestamp_s_;
  if (dt <= 1e-6) {
    return std::nullopt;
  }

  if (prev_points_.size() < static_cast<std::size_t>(config_.flow.min_features)) {
    prev_points_ = detect_features(prev_frame_);
    if (prev_points_.size() < 4) {
      bootstrap(frame_work, timestamp_s, frame_scale);
      return std::nullopt;
    }
  }

  std::vector<cv::Point2f> next_points;
  std::vector<uchar> status_fw;
  std::vector<float> error_fw;
  cv::calcOpticalFlowPyrLK(prev_frame_, frame_work, prev_points_, next_points, status_fw, error_fw, cv::Size(21, 21), 3);
  if (next_points.empty() || status_fw.empty()) {
    bootstrap(frame_work, timestamp_s, frame_scale);
    return std::nullopt;
  }

  std::vector<cv::Point2f> back_points;
  std::vector<uchar> status_bw;
  std::vector<float> error_bw;
  cv::calcOpticalFlowPyrLK(frame_work, prev_frame_, next_points, back_points, status_bw, error_bw, cv::Size(21, 21), 3);
  if (back_points.empty() || status_bw.empty()) {
    bootstrap(frame_work, timestamp_s, frame_scale);
    return std::nullopt;
  }

  std::vector<cv::Point2f> prev_valid;
  std::vector<cv::Point2f> next_valid;
  const double work_to_original_px = 1.0 / std::max(frame_scale, 1e-6);
  const std::size_t count =
    std::min({prev_points_.size(), next_points.size(), back_points.size(), status_fw.size(), status_bw.size()});
  for (std::size_t i = 0; i < count; ++i) {
    const double dx = back_points[i].x - prev_points_[i].x;
    const double dy = back_points[i].y - prev_points_[i].y;
    const double bidirectional_error = std::hypot(dx, dy) * work_to_original_px;
    if (status_fw[i] == 1 && status_bw[i] == 1 && bidirectional_error < config_.flow.max_bidirectional_error_px) {
      prev_valid.push_back(prev_points_[i]);
      next_valid.push_back(next_points[i]);
    }
  }

  if (prev_valid.size() < static_cast<std::size_t>(config_.flow.min_features)) {
    bootstrap(frame_work, timestamp_s, frame_scale);
    return std::nullopt;
  }

  std::vector<float> dx_values;
  std::vector<float> dy_values;
  dx_values.reserve(prev_valid.size());
  dy_values.reserve(prev_valid.size());
  for (std::size_t i = 0; i < prev_valid.size(); ++i) {
    dx_values.push_back(next_valid[i].x - prev_valid[i].x);
    dy_values.push_back(next_valid[i].y - prev_valid[i].y);
  }
  auto median = [](std::vector<float> values) {
      const auto mid = values.begin() + static_cast<std::ptrdiff_t>(values.size() / 2);
      std::nth_element(values.begin(), mid, values.end());
      return *mid;
    };
  const cv::Point2f flow_px_work(median(dx_values), median(dy_values));
  const cv::Point2f flow_px(
    static_cast<float>(static_cast<double>(flow_px_work.x) * work_to_original_px),
    static_cast<float>(static_cast<double>(flow_px_work.y) * work_to_original_px));
  const Vector2 velocity_xy_enu_mps = pixel_flow_to_velocity_enu(flow_px, dt, altitude_m, yaw_rad);

  const double feature_scale = std::clamp(config_.flow.min_features / std::max(static_cast<double>(prev_valid.size()), 1.0), 0.5, 4.0);
  const double sigma = config_.flow.noise_std_mps * feature_scale;
  Matrix2 covariance = Matrix2::Zero();
  covariance(0, 0) = sigma * sigma;
  covariance(1, 1) = sigma * sigma;

  frame_work.copyTo(prev_frame_);
  prev_timestamp_s_ = timestamp_s;
  prev_frame_scale_ = frame_scale;
  prev_points_ = next_valid;

  return FlowMeasurement{timestamp_s, velocity_xy_enu_mps, covariance, "optical_flow", static_cast<int>(prev_valid.size())};
}

cv::Mat OpticalFlowLayer::prepare_frame(const cv::Mat & frame_gray, double & scale)
{
  scale = 1.0;
  const int max_side_px = config_.flow.max_side_px;
  if (max_side_px <= 0) {
    return frame_gray;
  }
  const int max_side = std::max(frame_gray.cols, frame_gray.rows);
  if (max_side <= max_side_px) {
    return frame_gray;
  }
  scale = static_cast<double>(max_side_px) / static_cast<double>(max_side);
  const int resized_cols = std::max(1, static_cast<int>(std::round(static_cast<double>(frame_gray.cols) * scale)));
  const int resized_rows = std::max(1, static_cast<int>(std::round(static_cast<double>(frame_gray.rows) * scale)));
  cv::resize(frame_gray, resized_frame_, cv::Size(resized_cols, resized_rows), 0.0, 0.0, cv::INTER_AREA);
  return resized_frame_;
}

void OpticalFlowLayer::bootstrap(const cv::Mat & frame_gray, double timestamp_s, double scale)
{
  frame_gray.copyTo(prev_frame_);
  prev_timestamp_s_ = timestamp_s;
  prev_frame_scale_ = scale;
  prev_points_ = detect_features(frame_gray);
}

std::vector<cv::Point2f> OpticalFlowLayer::detect_features(const cv::Mat & frame_gray) const
{
  std::vector<cv::Point2f> points;
  cv::goodFeaturesToTrack(frame_gray, points, 800, 0.01, 6.0, cv::Mat(), 7, false);
  return points;
}

Vector2 OpticalFlowLayer::pixel_flow_to_velocity_enu(const cv::Point2f & flow_px, double dt, double altitude_m, double yaw_rad) const
{
  const double fx = camera_matrix_(0, 0);
  const double fy = camera_matrix_(1, 1);
  const double alt = std::max(1.0, std::abs(altitude_m));
  const double vx_body = static_cast<double>(flow_px.y) * alt / std::max(fy, 1e-6) / dt;
  const double vy_body = -static_cast<double>(flow_px.x) * alt / std::max(fx, 1e-6) / dt;
  const double c = std::cos(yaw_rad);
  const double s = std::sin(yaw_rad);
  return Vector2(c * vx_body - s * vy_body, s * vx_body + c * vy_body);
}

}  // namespace advanced_localization
